share/ipld/proof.go

```diff
@@ -0,0 +1,76 @@
+package ipld
+
+import (
+	"math"
+	"sort"
+	"sync"
+)
+
+// Proof contains information required for Leaves inclusion validation.
+type Proof struct {
+	Nodes      [][]byte
+	Start, End int
+}
+
+// proofCollector collects proof nodes with corresponding indexes and provides Nodes array that
+// could be used to construct nmt.Proof for shares inclusion validation.
+type proofCollector struct {
+	leftProofs, rightProofs []nodeWithIdx
+	leftLk, rightLk         sync.Mutex
+}
+
+type nodeWithIdx struct {
+	idx  int
+	node []byte
+}
+
+func newProofCollector(maxShares int) *proofCollector {
+	// maximum possible amount of required proofs from each side is equal to tree height.
+	height := int(math.Log2(float64(maxShares)))
+	return &proofCollector{
+		leftProofs:  make([]nodeWithIdx, 0, height),
+		rightProofs: make([]nodeWithIdx, 0, height),
+	}
+}
+
+func (c *proofCollector) addLeft(idx int, node []byte) {
+	c.leftLk.Lock()
+	defer c.leftLk.Unlock()
+	c.leftProofs = append(c.leftProofs, nodeWithIdx{
+		idx:  idx,
+		node: node,
+	})
+}
+
+func (c *proofCollector) addRight(idx int, node []byte) {
+	c.rightLk.Lock()
+	defer c.rightLk.Unlock()
```

Comment:
Can we try to keep `GetShareByNamespace` lock-free? These are now the only locks there. What if we treat `idx` as a slice index?

Reply:
Good suggestion, thank you. I've realised there is no need for locks, because there is actually no races. Updated.